### PR TITLE
Add E0124 and E0277 error explanation

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of #24407.
cc @michaelsproul
r? @Manishearth

### CODE REVIEW

File: src/librustc/diagnostics.rs [hunk 1]
```diff
@@ -749,6 +749,54 @@ for v in &vs {
 ```
 "##,
 
+E0277: r##"
+You tried to called a trait function on a type which doesn't implement this
```
Comment (Manishearth) [inline src/librustc/diagnostics.rs]: You tried to use a type which doesn't implement some trait in a place which expected that trait.

File: src/librustc/diagnostics.rs [hunk 2]
```diff
@@ -749,6 +749,54 @@ for v in &vs {
...
+
+fn main() {
+    // we now call the method with the i32 types, which doesn't implement
+    // the Foo trait
+    some_func(5i32); // error: the trait `Foo` is not implemented for the
+                     //     type `i32`
+}
+```
+
+In order to fix this error, please check the type you're using does implement
```
Comment (Manishearth) [inline src/librustc/diagnostics.rs]: check the type -> verify that the type

File: src/librustc_typeck/diagnostics.rs [hunk 3]
```diff
@@ -1347,6 +1347,27 @@ static BAR: _ = "test"; // error, explicitly write out the type instead
...
+example:
+
+```
+struct Foo {
+    field1: i32,
+    field1: i32 // error: field is already declared
+}
+```
+
+Please check you didn't mispelled one field. Example:
```
Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: Please verify that the field names have been correctly spelled

File: src/librustc/diagnostics.rs [hunk 4]
```diff
@@ -749,6 +749,54 @@ for v in &vs {
...
+}
+
+// we now declare a function which takes an object with Foo trait implemented
+// as parameter
+fn some_func<T: Foo>(foo: T) {
+    foo.bar();
+}
+
+fn main() {
+    // we now call the method with the i32 types, which doesn't implement
```
Comment (apasel422) [inline src/librustc/diagnostics.rs]: s/types/type/

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by Manishearth** [comment]
r=me post squash

**Comment by GuillaumeGomez** [comment]
Squashed.

**Pull request merged**